ipykernel/kernelbase.py

```diff
@@ -984,6 +1066,63 @@ async def usage_request(self, socket, ident, parent):
     async def do_debug_request(self, msg):
         raise NotImplementedError
 
+    # ---------------------------------------------------------------------------
+    # Subshell control message handlers
+    # ---------------------------------------------------------------------------
+
+    async def create_subshell_request(self, socket, ident, parent):
+        if not self.session:
+            return
+        if not self._supports_kernel_subshells():
+            self.log.error("Subshells are not supported by this kernel")
+            return
+
+        # This should only be called in the control thread if it exists.
+        # Request is passed to shell channel thread to process.
+        other_socket = self.shell_channel_thread.manager.get_control_other_socket()
+        await other_socket.send_json({"type": "create"})
+        reply = await other_socket.recv_json()
+
+        self.session.send(socket, "create_subshell_reply", reply, parent, ident)
+
+    async def delete_subshell_request(self, socket, ident, parent):
+        if not self.session:
+            return
+        if not self._supports_kernel_subshells():
+            self.log.error("KERNEL SUBSHELLS NOT SUPPORTED")
+            return
```

Comment:
Should that be a hard error ?

I have the feeling that this should never happen or if it is should be handle/silence at higher level.

(Also I dont' think uppercase necessary)

Reply:
This is based on existing control message handling, if we receive a message that we don't recognise we log an error message and then carry on without considering it an error situation:
https://github.com/ipython/ipykernel/blob/0fa543936d3f9ee69f6b9f8bd5a316692fde9aaa/ipykernel/kernelbase.py#L277
and:
https://github.com/ipython/ipykernel/blob/0fa543936d3f9ee69f6b9f8bd5a316692fde9aaa/ipykernel/kernelbase.py#L291

But I suppose here the message type and contents are valid, but the kernel is not capable of doing what is requested. We could return a reply message with `{'status': 'error'}`? What to do here is not covered by the JEP.

Uppercase is certainly not necessary!

Reply:
I would prefer starting with a hard failure, worse case in a minor revision we ignore. The main reason is that raising where it was not previously have more chance to break downsrtream users and is easier to track than just ignoring anything. Returning status:error woudl be ok-ish I guess.

Reply:
To clarify, by 'hard failure' do you mean raise an exception?